src/utils/object_pool.hpp

```diff
@@ -0,0 +1,245 @@
+    std::cout << inuse_.size() << " used objects." << std::endl;
+  }
+
+  // This should be used with care since it can't generically be
+  // checked that the input object has the same size as other objects
+  // in the pool
+  void AddFreeObjectToPool(const T &in) { available_.push(in); }
+  void AddFreeObjectToPool(T &&in) { available_.emplace(in); }
+
+ private:
+  bool IsValid(const weak_t &in) const { return inuse_.count(in.key_); }
+
+  void ReferenceCountedFree(const weak_t &in) {
+    if (!IsValid(in)) return;
+    auto &pair = inuse_[in.key_];
+    --pair.second;
+    if (pair.second <= 0) {
+      available_.push(pair.first);
+      inuse_.erase(in.key_);
+    }
+  }
+
+  void Free(const weak_t &in) {
+    if (!IsValid(in)) return;
+    available_.push(inuse_[in.key_].first);
+    inuse_.erase(in.key_);
+  }
+
+  void AddCount(const weak_t &in) {
+    if (!IsValid(in)) throw 1;
+    ++inuse_[in.key_].second;
+  }
+};
+
+// Lightly wraps a view of type T and holds a key that
+// that can be used to query if its storage
+// is in use or if it has been freed and allows
+// freeing of the storage.
+template <class T>
+struct ObjectPool<T>::weak_t : public T {
+  friend class ObjectPool;
+
+ protected:
+  template <class... ARGs>
+  KOKKOS_IMPL_HOST_FUNCTION static weak_t make(int key, ARGs &&...args) {
+    weak_t out(std::forward<ARGs>(args)...);
+    out.key_ = key;
+    return out;
+  }
+
+ public:
+  template <class... Ts>
+  KOKKOS_IMPL_HOST_FUNCTION explicit weak_t(Ts &&...args)
+      : T(std::forward<Ts>(args)...), key_(default_key_) {}
+
+  KOKKOS_IMPL_HOST_FUNCTION
+  inline void Free() { (*pool_).Free(*this); }
+
+  KOKKOS_IMPL_HOST_FUNCTION
+  inline bool IsValid() {
+    if (key_ == default_key_ || pool_ == nullptr) return false;
+    return (*pool_).IsValid(*this);
+  }
+
+  KOKKOS_IMPL_HOST_FUNCTION
+  inline KEY_T GetKey() const { return key_; }
+
+  KOKKOS_DEFAULTED_FUNCTION
+  ~weak_t() = default;
+
+  KOKKOS_DEFAULTED_FUNCTION
+  weak_t() = default;
+
+  KOKKOS_DEFAULTED_FUNCTION
+  weak_t(const weak_t &) = default;
+
+  KOKKOS_DEFAULTED_FUNCTION
+  weak_t(weak_t &&) = default;
+
+  KOKKOS_DEFAULTED_FUNCTION
+  weak_t &operator=(const weak_t &) = default;
+
+  KOKKOS_DEFAULTED_FUNCTION
+  weak_t &operator=(weak_t &&) = default;
+
+  // Allow this to point at an unmanaged object of type T
+  KOKKOS_FUNCTION
+  weak_t &operator=(const T &in) {
+    T::operator=(in);
+    return *this;
+  }
+
+ protected:
+  KEY_T key_ = 0;
+  ObjectPool *pool_ = nullptr;
+};
+
+// Reference counted version of pool member that has ownership over a resource
+// and sends it back to the pool when its destructor is called and it is the
+// last owner that holds that resource. Cannot be on device
+template <class T>
+class ObjectPool<T>::owner_t : public ObjectPool<T>::weak_t {
+ public:
+  KOKKOS_DEFAULTED_FUNCTION
+  owner_t() = default;
+
+  KOKKOS_FUNCTION
+  ~owner_t() noexcept {
+#ifndef __CUDA_ARCH__ // host code
```

Comment:
What does this mean for HIP, SYCL and other backend?

Reply:
I think this is not ideal and may break for other backends, but I do not know a `Kokkos` based way to call something depending on whether the function is being called from host or device.

Reply:
If there are not objections to bump kokkos to 3.6.01 after the next release and prior to merging this, which, actually will be required in order to properly run on Frontier and should  have no side effects on the current code base, then there's a new `KOKKOS_IF_ON_{HOST/DEVICE}` macro.

Reply:
switched